File: baekjoon/python/4673/4673.py
# import sys
# sys.stdin = open('4673_input.txt','r')

# 10,000보다 작거나 같은 셀프 넘버를 한 줄에 하나씩 증가하는 순서로 출력한다.
'''
셀프넘버
1부터 dp 조지기?
recursive로 채워
1+1 = 2
2+2 = 4
3+3 = 6
4+4 = 8
5+5 = 10
6+6 = 12
..
10+1+0=11

1 -> 1+1 -> 2 -> 2+2 -> 4 -> 4+4 -> 8 -> 8+8 -> 16 -> 16+1+6 -> 23 -> 23+2+3 -> 28

'''

dp = [0]*10001


def posnum(num):

    sum = 0
    for i in range(4, -1, -1):
        sum += int(num/(10**i))
        num %= 10**i
    return sum


def recursive(num):  # dp가 0이라면 계속 진행 / 10000보다 작다면 계속 진행
    num += posnum(num)
    if num > 10000:  # 10000으로 수정
        return 0
    if dp[num] == 0:
        dp[num] = 1
        return recursive(num)
    else:
        return 0


for i in range(10001):
    recursive(i)

for i, num in enumerate(dp):
    if num == 0:
        print(i)

File: baekjoon/cpp/4673/4673.cpp
#include <bits/stdc++.h>
using namespace std;

int dp[10001] = {
    0,
};

int pos_sum(int n)
{
    /*
    12345
    1+2+3+4+5
    한자리씩 줄여가기
    */
    int sum = 0;
    while (n != 0)
    {
        sum += n % 10;
        n /= 10;
    }
    return sum;
}

int self(int n)
{
    // n을 쪼개기
    /*
    1+1=2
    2+2=4
    11 -> 11 + 1 + 1
    1234 -> 1234 + 1+2+3+4
    */
    n += pos_sum(n);
    if (n > 10000) // 10000넘는 un-self number
        return 0;
    else if (dp[n] == 0) // 본적없는 un-self number
    {
        dp[n] = 1;
        return self(n);
    }
    else // 이미 해본 거
    {
        return 0;
    }
}
int main()
{
    std::ios_base::sync_with_stdio(false);
    freopen("4673_input.txt", "r", stdin);

    for (int i = 1; i <= 10000; i++)
    {
        self(i);
    }
    for (int i = 1; i <= 10000; i++)
    {
        if (!dp[i])
            cout << i << endl;
    }

    return 0;
}
